File: packages/asdl-core/src/asdl_core/sessions/source.py
"""Session source interface."""

from __future__ import annotations

from abc import ABC, abstractmethod

from asdl_core.sessions.types import SessionQuery, SessionQueryResult, SessionSourceInfo


class SessionSource(ABC):
    """Readable source of harness session facts."""

    @property
    @abstractmethod
    def source_info(self) -> SessionSourceInfo:
        """Return adapter identity without filesystem or subprocess work."""

    @abstractmethod
    def query(self, query: SessionQuery) -> SessionQueryResult:
        """Return sessions and non-fatal warnings for the query."""

File: ts/packages/capabilities/retros/src/sessions/source.ts
import type { SessionQuery, SessionQueryResult, SessionSourceInfo } from "./types.ts";

export interface SessionSource {
	readonly sourceInfo: SessionSourceInfo;
	query(query: SessionQuery): Promise<SessionQueryResult>;
}

export function limitSessions<TSession>(
	sessions: readonly TSession[],
	maxSessions: number,
): TSession[] {
	if (maxSessions <= 0) {
		return [];
	}
	return sessions.slice(0, maxSessions);
}
